karabo_data/geometry2.py

```diff
@@ -205,6 +217,141 @@ def inspect(self, frontview=True):
+        return rigid_string
+
+    def write_crystfel_geom(self, filename, *,
+                            data_path='/entry_1/instrument_1/detector_1/data',
+                            mask_path=None, dims=('frame', 'modno', 'ss', 'fs'),
+                            adu_per_ev=None, clen=None, photon_energy=None):
+        """Write this geometry to a CrystFEL format (.geom) geometry file.
+        Parameters
+        ----------
+        filename : str
+        filename of the geometry file
+
+        data_path  : str, default : /entry_1/instrument_1/detector_1/data
+            path to the group that contains the data array in the hdf5 file
+        mask_path : str
+            path to the group that contains the mask array in the hdf5 file
+
+        dims : collection, default : ('frame', 'modno', 'ss', 'fs')
+            Dimensions of the data. Extra dimensions, except for the defaults,
+            should be added by their index, e.g.
+            ('frame', 'modno', 0, 'ss', 'fs') for raw data.
+            Note: the dimensions must contain frame, modno, ss, fs.
+
+        adu_per_ev : float
+            ADU (analog digital units) per electron volt for the considered
+            detector.
+
+        clen : float
+            Distance between sample and detector in meters
+
+        photon_energy
+            Beam wave length in eV
+        """
+        from . import __version__
+
+        if adu_per_ev is None:
+            adu_per_ev_str = '; adu_per_eV = SET ME'
+            # TODO: adu_per_ev should be fixed for each detector, we should 
+            #       find out the values and set them.
+        else:
+            adu_per_ev_str = 'adu_per_eV = {}'.format(adu_per_ev)
+
+        if clen is None:
+            clen_str = '; clen = SET ME'
+        else:
+            clen_str = 'clen = {}'.format(clen)
+
+        if photon_energy is None:
+            photon_energy_str = '; photon_energy = SET ME'
+        else:
+            photon_energy_str = 'photon_energy = {}'.format(photon_energy)
+
+        # Get the frame dimension
+        tile_dims = {}
+
+        for nn, dim_name in enumerate(dims):
+            if dim_name == 'frame':
+                frame_dim = 'dim{} = %'.format(nn)
+            else:
+                tile_dims[nn] = dim_name
+        if frame_dim is None:
+            raise ValueError ('No frame dimension given')
+
+        panel_chunks = []
+        for p, module in enumerate(self.modules):
+            for a, fragment in enumerate(module):
+                ss_dims, fs_dims = self._tile_dims(a)
+                panel_chunks.append(fragment.to_crystfel_geom(p,
+                                                              a,
+                                                              ss_dims,
+                                                              fs_dims,
+                                                              tile_dims))
+        pix_size = self.pixel_size * 1e6 # Convert pixels size to microns
```

Comment:
While doing a final read-through, I spotted that this isn't right. CrystFEL expresses resolution in pixels per metre. I'm adding a fix for that.